Added: System.Text.Json serialization support

DSilence opened the pull request:
This is an initial implementation of System.Text.Json support. The code is mostly a copy of Json.Net serialization routine, since the API is very similar. There is a lot of potential for additional performance improvements, but for now I think compatibility is more important.

TODO:
- [ ] Fix unit tests. 
Some of them are failing due to `Utf8JsonWriter.WriteStringValue` method escaping the forward slash symbol. As of right now, I couldn't find a way to override this, and the API allowing it doesn't seem to be a part of preview 7 release (https://github.com/dotnet/corefx/issues/37192). Will get back to it at a later time.

Looking forward to comments/suggestions!

jskeet commented:
I'll get on this when I can, but it may be some time. It might be worth merging this into a separate branch as a work in progress until the tests pass - what do you think?

DSilence commented:
Sounds like a good idea!

jskeet commented:
Okay, I've created the new branch and retargeted this. Anything else you want to do now, or shall I just merge into this branch and review when I get a chance?

DSilence commented:
Thanks!
There seems to be extra null safety built into the JsonReader itself. So some of the code in the base converter might not be required. I'd like to take another look through the codebase and double check those cases before the merge.

jskeet commented:
Fine by me - just leave another comment when you'd like to merge :)

DSilence commented:
I've cleaned up the code a bit. Below is the list of changes:
1) The framework handles the nullable value types itself. There is no need for extra voodoo in base converter. Basically, the base converter now only exists to return proper exception type on deserializing if anything goes wrong.
2) The nullable type handling will, however, no longer treat empty string as null values. So this case will no longer be supported. I've removed the tests related to that. 
3) The exception which is thrown when deserialization fails was changed to `System.Text.Json.JsonException` as it seems more appropriate than a more generic SerializationException I've used before.

jskeet commented:
I'll have a brief look when I get back from holiday, and then merge into the branch.

DSilence commented:
Sure thing, thanks. Let me know if some adjustments will be required.
